pkg/provider/ngrok/provider.go

```diff
@@ -0,0 +1,183 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+	"github.com/external-secrets/external-secrets/pkg/utils/resolvers"
+	"github.com/ngrok/ngrok-api-go/v7"
+	"github.com/ngrok/ngrok-api-go/v7/secrets"
+	"github.com/ngrok/ngrok-api-go/v7/vaults"
+)
+
+var (
+	defaultAPIURL = "https://api.ngrok.com"
+	userAgent     = "external-secrets"
+
+	errClusterStoreRequiresNamespace = errors.New("cluster store requires namespace")
+	errInvalidStore                  = errors.New("invalid store")
+	errInvalidStoreSpec              = errors.New("invalid store spec")
+	errInvalidStoreProv              = errors.New("invalid store provider")
+	errInvalidNgrokProv              = errors.New("invalid ngrok provider")
+	errInvalidAuthAPIKeyRequired     = errors.New("ngrok provider auth APIKey is required")
+	errInvalidAPIURL                 = errors.New("invalid API URL")
+	errMissingVaultName              = errors.New("ngrok provider vault name is required")
+)
+
+type vaultClientFactory func(cfg *ngrok.ClientConfig) VaultClient
+type secretsClientFactory func(cfg *ngrok.ClientConfig) SecretsClient
+
+var getVaultsClient vaultClientFactory = func(cfg *ngrok.ClientConfig) VaultClient {
+	return vaults.NewClient(cfg)
+}
+
+var getSecretsClient secretsClientFactory = func(cfg *ngrok.ClientConfig) SecretsClient {
+	return secrets.NewClient(cfg)
+}
+
+type Provider struct{}
+
+// Capabilities returns the provider supported capabilities (ReadOnly, WriteOnly, ReadWrite). Currently,
+// ngrok only supports WriteOnly capabilities.
+func (p *Provider) Capabilities() esv1.SecretStoreCapabilities {
+	return esv1.SecretStoreWriteOnly
+}
+
+// NewClient implements the Client interface.
+func (p *Provider) NewClient(ctx context.Context, store esv1.GenericStore, kubeClient kubeClient.Client, namespace string) (esv1.SecretsClient, error) {
+	cfg, err := getConfig(store)
+	if err != nil {
+		return nil, err
+	}
+
+	if store.GetKind() == esv1.ClusterSecretStoreKind && doesConfigDependOnNamespace(cfg) {
+		return nil, errClusterStoreRequiresNamespace
+	}
+
+	apiKey, err := loadAPIKeySecret(ctx, cfg.Auth.APIKey, kubeClient, store.GetKind(), namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	clientConfig := ngrok.NewClientConfig(
+		apiKey,
+		ngrok.WithBaseURL(cfg.APIURL),
+		ngrok.WithUserAgent(userAgent),
+	)
+
+	vaultClient := getVaultsClient(clientConfig)
+	secretsClient := getSecretsClient(clientConfig)
+
+	var vault *ngrok.Vault
+	vaultIter := vaultClient.List(nil)
+	for vaultIter.Next(ctx) {
```

Comment:
Okay, so I looked at the code in ngrok and I'm pretty sure that we need a context with a timeout for all iterator calls. If ngrok becomes unresponsive this could hang for a long time since the Do in ngrok doesn't have a timeout either.

WDYT?

Reply:
With the exception of the validation on the provider, I've added a defaultListTimeout of 1 minute and updated the  other List calls to use it in https://github.com/external-secrets/external-secrets/pull/5160/commits/a9eefb0edff85f520abf23e764f51c84d4cd383e. If you think that is too long or short, I can change it.